Support for the ZHash algo fork

renuzit opened the pull request:
This commit does a couple things:
* Add support for ZHash (EH144,5)
* Update protocol to 770004
* Bumps to v1.2.0
* Update testnet servers

NickRHill committed:
WIP add variable EH params

NickRHill committed:
WIP add chainparams and notes

NickRHill committed:
WIP add block height lookup to EH validation

NickRHill committed:
remove height check from context-insensitive code

NickRHill committed:
add overlapping EH parameter verification

NickRHill committed:
clean up test parameters

NickRHill committed:
bork mainnet

NickRHill committed:
bork mainnet

NickRHill committed:
correct EH pers string

NickRHill committed:
add testnet seed

NickRHill committed:
Enable mainnet, add checkpoints

NickRHill committed:
update versions

NickRHill committed:
testnet mining requires peers

renuzit committed:
Merge remote-tracking branch 'btcz/master'

renuzit committed:
Bump version number to v1.2.0

renuzit committed:
Update version

renuzit committed:
Support only version 770002 and above

renuzit commented on src/chainparams.cpp:
```diff
@@ -61,10 +61,11 @@ class CMainParams : public CChainParams {
-        nEquihashK = K;
+        eh_epoch_1 = eh200_9;
+        eh_epoch_2 = eh144_5;
+        eh_epoch_1_endblock = 170000;
```
If we decide to keep the overlap period, I think `eh_epoch_1_endblock` should be no more than 161152.

renuzit replied:
The more I think about this, the more I would like to see the overlap period only be 10 blocks which would be a 25 minute window.

NickRHill replied:
Since the difficulty adjustment is not sensitive to the algorithm, then I think there should be no overlap.

renuzit replied:
I think a little overlap makes sense. It's what we've tested and I would feel more comfortable with a 10 block overlap than one with 0.

renuzit commented on src/main.cpp:
```diff
@@ -3082,7 +3082,28 @@ bool ContextualCheckBlockHeader(const CBlockHeader& block, CValidationState& sta
+        return state.DoS(100,error("ContextualCheckBlockHeader: Equihash solution size %d for block %d does not match a valid length",nSolSize, nHeight),
+           REJECT_INVALID,"bad-equihash-solution-size");
+    }
 
```
Need to drop the difficulty for in `GetNextWorkRequired` for ZHash.

renuzit replied:
@NickRHill What do you think about this?

```diff
diff --git a/src/pow.cpp b/src/pow.cpp
index e77ff44..1e40327 100644
--- a/src/pow.cpp
+++ b/src/pow.cpp
@@ -22,12 +22,19 @@
 
 unsigned int GetNextWorkRequired(const CBlockIndex* pindexLast, const CBlockHeader *pblock, const Consensus::Params& params)
 {
+    const CChainParams& chainParams = Params();
     unsigned int nProofOfWorkLimit = UintToArith256(params.powLimit).GetCompact();
 
     // Genesis block
     if (pindexLast == NULL)
         return nProofOfWorkLimit;
 
+    // Reset the difficulty after the algo fork
+    if (pindexLast->nHeight > chainParams.eh_epoch_1_end() - 1
+      && pindexLast->nHeight < chainParams.eh_epoch_1_end() + params.nPowAveragingWindow) {
+      return nProofOfWorkLimit;
+    }
+
     // Find the first block in the averaging interval
     const CBlockIndex* pindexFirst = pindexLast;
     arith_uint256 bnTot {0};
```

NickRHill replied:
Looks like that will work.

renuzit replied:
pushed

renuzit commented on src/main.cpp:
```diff
@@ -3082,7 +3082,28 @@ bool ContextualCheckBlockHeader(const CBlockHeader& block, CValidationState& sta
 
```
I feel like something is missing in `CheckBlockHeader`

NickRHill replied:
Regarding the difficulty or some other change?

renuzit replied:
Nevermind.

NickRHill reviewed (changes requested):
Before merging into public main, in version.h set:
static const int PROTOCOL_VERSION = 770004;
static const int MIN_PEER_PROTO_VERSION = 770004;
And in chainparams.cpp
eh_epoch_1_endblock = 160000;

renuzit commented on src/pow/tromp/equi.h:
```diff
@@ -17,11 +17,11 @@ typedef unsigned char uchar;
 
 #ifndef WN
-#define WN	144
+#define WN	200
```
This change was a mistake. We need to use defaults here, because the solution size is set programatically. In other words, old blocks are still 200,9.

renuzit committed:
Reset the difficulty after the algo fork

Once we switch over to the new algorithm the current difficulty
setting should be a lot higher than what it should be. This patch
resets the difficulty back to 0 for 13 blocks so that we can properly
determine what the new difficulty should be.

renuzit commented:
@NickRHill
We are currently running Protocal Version 770002. PROTOCOL_VERSION = 770003 would be the new version. MIN_PEER_PROTO_VERSION should be set to 770003

Changing eh_epoch_1_endblock to 160010 for a small overlap as an overlap period was in our testing.

renuzit committed:
Fixed a few issues found in testing

While testing we discovered that the miner did not start mining on
the right block, so updated the code to fix the off-by-one error.
Also, included a few nit pick cleanups

renuzit committed:
Bumped the protocol version to 770004

renuzit committed:
Responding to PR feedback

renuzit committed:
Bump version to 770004

NickRHill commented:
I think we also need to crop seeds we don't have contact with and who are not resolving.

NickRHill committed:
update DNS seeds
